src/uint/boxed/shl.rs

```diff
@@ -1,108 +1,127 @@
 //! [`BoxedUint`] bitwise left shift operations.
 
-use crate::{BoxedUint, CtChoice, Limb, Word};
+use crate::{BoxedUint, Limb};
 use core::ops::{Shl, ShlAssign};
 use subtle::{Choice, ConstantTimeLess};
 
 impl BoxedUint {
     /// Computes `self << shift`.
-    /// Returns zero if `shift >= Self::BITS`.
-    pub fn shl(&self, shift: u32) -> Self {
+    /// Returns `None` if `shift >= Self::BITS`.
+    pub fn shl(&self, shift: u32) -> (Self, Choice) {
```

Comment:
This could perhaps be:

```suggestion
    pub fn shl(&self, shift: u32) -> CtOption<Self> {
```

Reply:
I mentioned that in the description - the return type is not `CtOption` here since most of its methods need the type to be `ConditionallySelectable`, which `BoxedUint` isn't, so it would be pretty useless

Reply:
There are some fake `CtOption` combinator methods defined on `BoxedUint`:

- https://docs.rs/crypto-bigint/0.6.0-pre.1/crypto_bigint/struct.BoxedUint.html#method.conditional_and_then
- https://docs.rs/crypto-bigint/0.6.0-pre.1/crypto_bigint/struct.BoxedUint.html#method.conditional_map

Though ideally this would get solved upstream on `subtle`: https://github.com/dalek-cryptography/subtle/pull/118

cc @isislovecruft

Reply:
Also there are other methods for the `Boxed*` types already using `CtOption`, so we should probably choose a convention for `BoxedUint`/`BoxedResidue` and stick to it

Reply:
The problem is that these methods *return* a `CtOption`, but there are no methods that *take* a `CtOption`, so if you end up with one, there is not much you can do besides unwrapping it in variable time. In one place in the codebase I need an analogue of `unwrap_or()`, how would I implement that?

Reply:
Yeah, really this would ideally be solved upstream in `subtle`.

I think the `(Self, CtChoice)` tuples are a bit error-prone since they still expose the inner value even if `CtChoice` is false. If you wanted to define a similar analogue to `CtOption` which works for the `const` cases and `CtChoice`, that seems good to me

Reply:
> I think the (Self, CtChoice) tuples are a bit error-prone since they still expose the inner value even if CtChoice is false.

In the specific case of shifts we can return zero in case of overflow, which is a perfectly logical and often useful value (so sometimes you can just ignore the overflow flag). It is documented for `Uint` shifts, but not for `BoxedUint`, so if we document it there too, it should be relatively ok. 

> If you wanted to define a similar analogue to CtOption which works for the const cases and CtChoice, that seems good to me

I thought about it too. It would also be useful to distinguish the cases where `CtChoice` denotes an overflow, a carry, or that the result exists at all (like in inversions).

Reply:
If `BoxedUint` were to use this new const analogue of `CtOption` it would not be consisted with the rest of `BoxedUint` API that uses `subtle`, so I think for now a tuple is better.

Reply:
Ok, let's go with this for now, though we should perhaps make a tracking issue to make all of the APIs consistent

Reply:
Filed #422 for the `CtOption` equivalent